rfc/system/0000-caching-of-offloaded-objects.md

```diff
@@ -0,0 +1,161 @@
+    string hash = 4;
+}
+```
+
+### Flytekit 
+
+The crux of the mechanics in flytekit revolves around how to expose enough flexibility to allow for arbitrary hash functions to be used, while at the same time providing enough information to flytekit. We propose the use of a `HashMethod` metadata object used in annotated return types. The idea being that during the process of converting from a python value to a literal we apply that hash method and set it in the literal.
+
+We'll lean on [`typing.Annotated`](https://docs.python.org/3/library/typing.html#typing.Annotated) to check the types and annotations. 
+
+Since we cannot assume any hashing characteristics of the hash functions (since we are not talking about perfect hashing in the general case), the `HashMethod` will expose a simple interface, composed of a callable of type `Callable[[T], str]`. That said, we will provide plenty of examples to showcase the flexibility of this approach.
+
+### Cache key calculation
+
+As mentioned above, during the regular cache key calculation, in both the local and the remote cases, flyte takes the input literal map as part of the key. Going forward we will allow for specific literals in that map to use the overridden hash as the representation of that literal.
+
+### Other clients
+
+Although nothing prevents the adoption of this feature in other clients, flytekit-java and other clients are outside of the scope of this RFC.
+
+## 4 Metrics & Dashboards
+
+N/A?
+
+## 5 Drawbacks
+
+Although this feature does *not* impact the other aspects of how the cache works, for example,  changes to the version or the signature of the task invalidate the cache entries, it still might cause confusion since from the perspective of a task execution we will not be able to tell if an object will have its hash overridden. 
+
+## 6 Alternatives
+
+A few options were discussed in https://github.com/flyteorg/flyte/issues/1581:
+
+**Expose a `hash` method in Type Transformers and annotate the task in case the type transformer does not contain a hash function:** The major drawbacks of this alternative are two-fold: 
+    (1) There's no way to opt-out, i.e. all objects produced by that Type Transformer will be cached.
+    (2) The UX is a bit clunky in the case of user-defined hash functions for two reasons:
+        - the return type does not contain any indication that the object is being cached, i.e. we'd have to look in the parameters set in the @task decorator
+
+**Offload the hash calculation to an external system like s3:** The idea would be to rely on etags produced by the act of storing an object in s3. This turns out to be a chicken-and-egg problem as we need to produce etags in order to compare if that tag was already produced. Also, even if we were able to overcome that hurdle, handing this computation to a blackbox system like s3 risks making cache entries unrecoverable, since there is no guarantee that the algorithm s3 uses to compute etags will not change.
```

Comment:
so in the proposed method, users can use etags or not? By first glance it seems not possible, but it depends on how the hashing function is invoked. We could change the signature to include the S3 / remote reference and then the hashing function could use that to capture the key.
Also in this case we could upload first and then call the hashing function?

Reply:
etags are associated with a specific resource (think of the tuple `(uri, contents)`, although I'm oversimplifying since even certain characteristics of the client might influence the generation of the etag), but the remote references we generate are not stable (in the sense that uri's can change). So unless I'm missing something, even if we had access to etags when we first upload an object, there's no way for us to have access to that specific resource in a second call so that we could check the etag on the original object.

To better illustrate this point, let's go back to the example in the doc:
```
@task
def foo(a: int, b: str) -> Annotated[pd.DataFrame, HashMethod(hash_pandas_dataframe_function) :
    df = pd.Dataframe(...)
    ...
    return df

@task(cached=True, version="1.0")
def bar(df: pd.Dataframe) -> int:
    ...
    
@workflow
def wf(a: int, b: str):
    df = foo(a=a, b=b)
    ...
    v = bar(df=df) 
```

Let's say that on the first execution of `foo` we decide to store the literal at s3://a/b/c with etag header set to 1234. The next time we execute `foo` we might decide to store this new literal at s3://aa/bb/cc, which would very likely be associated with a different etag.

The problem really is the fact that our process of producing the uri's [is random](https://github.com/flyteorg/flytekit/blob/master/flytekit/types/schema/types_pandas.py#L133), which renders etags incompatible (since they are associated with a single resource).